### PR TITLE
Update async retry mechanisms

### PR DESCRIPTION
**Pull request by katrogan**
# TL;DR
Add retries to notifications processor to handle underlying SQS connection failures.

In general, add user-config specified retries for initializing the initial AWS/GCP client but otherwise indefinitely retry in the case an open channel hiccups.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/lyft/flyte/issues/376

## Follow-up issue
_NA_

### CODE REVIEW

File: pkg/rpc/adminservice/base.go [hunk 1]
```diff
@@ -98,11 +98,18 @@ func NewAdminServer(kubeConfig, master string) *AdminService {
 	publisher := notifications.NewNotificationsPublisher(*configuration.ApplicationConfiguration().GetNotificationsConfig(), adminScope)
 	processor := notifications.NewNotificationsProcessor(*configuration.ApplicationConfiguration().GetNotificationsConfig(), adminScope)
 	go func() {
+		logger.Info(context.Background(), "Started processing notifications.")
 		err = processor.StartProcessing()
```
Comment (kumare3) [inline pkg/rpc/adminservice/base.go:102]: shouldnt start processing auto-handle connection loss and reconnect failures? Ofcourse it should log

Comment (katrogan) [reply pkg/rpc/adminservice/base.go]: sure, refactored @kumare3

**Comment by kumare3** [comment]
Looks good to me, just one thing, in case an error connection, should FlyteAdmin process not crash/exit?

**Comment by katrogan** [comment]
I'm not sure - is a running admin with partial functionality better or should we fail fast?

**Comment by kumare3** [comment]
> I'm not sure - is a running admin with partial functionality better or should we fail fast?

IMO, it could be a network partition or some other issue. If we fail fast, we might be able to recover on a different node? Also, since we have multiple replica's others should be fine.

On the other hand in sandbox mode we should probably not crash as users may not be able to figure out what the problem is. I pinged Ruslan. I guess if we do it for this, we should probably do it for all other cases?

**Comment by katrogan** [comment]
Updated so that we retry initializing the initial AWS/GCP client the number of times specified in the config but otherwise indefinitely retry in the case an open channel hiccups.

File: pkg/async/notifications/factory.go [hunk 2]
```diff
@@ -2,6 +2,9 @@ package notifications
 
 import (
 	"context"
+	"time"
+
+	"github.com/lyft/flyteadmin/pkg/async"
```
Comment (kumare3) [inline pkg/async/notifications/factory.go:7]: looks like a future flytestdlib entity

File: pkg/async/notifications/factory.go [hunk 3]
```diff
@@ -37,11 +40,18 @@ type EmailerConfig struct {
 	BaseURL     string
 }
 
-func GetEmailer(config runtimeInterfaces.NotificationsConfig, scope promutils.Scope) interfaces.Emailer {
+func GetEmailer(config runtimeInterfaces.NotificationsConfig, scope promutils.Scope,
+	reconnectAttempts int, reconnectDelay time.Duration) interfaces.Emailer {
 	switch config.Type {
 	case common.AWS:
 		awsConfig := aws.NewConfig().WithRegion(config.Region).WithMaxRetries(maxRetries)
```
Comment (kumare3) [inline pkg/async/notifications/factory.go:47]: it seems aws config already takes retries and delay?

Comment (katrogan) [reply pkg/async/notifications/factory.go:47]: derp, thanks for the catch. updated

File: pkg/async/notifications/implementations/processor.go [hunk 4]
```diff
@@ -38,7 +41,16 @@ type Processor struct {
 // Currently only email is the supported notification because slack and pagerduty both use
 // email client to trigger those notifications.
 // When Pagerduty and other notifications are supported, a publisher per type should be created.
-func (p *Processor) StartProcessing() error {
+func (p *Processor) StartProcessing() {
+	for {
+		logger.Warningf(context.Background(), "Starting notifications processor")
+		err := p.run()
+		logger.Errorf(context.Background(), "error with running processor err: [%v] ", err)
```
Comment (wild-endeavor) [inline pkg/async/notifications/implementations/processor.go:48]: should we log a metric here too?

File: pkg/async/schedule/aws/workflow_executor.go [hunk 5]
```diff
@@ -168,6 +171,15 @@ func (e *workflowExecutor) formulateExecutionCreateRequest(
 }
 
 func (e *workflowExecutor) Run() {
+	for {
+		logger.Warningf(context.Background(), "Starting workflow executor")
+		err := e.run()
```
Comment (wild-endeavor) [inline pkg/async/schedule/aws/workflow_executor.go:176]: and here

File: pkg/runtime/interfaces/application_configuration.go [hunk 6]
```diff
@@ -149,6 +149,10 @@ type NotificationsConfig struct {
 	NotificationsPublisherConfig NotificationsPublisherConfig `json:"publisher"`
 	NotificationsProcessorConfig NotificationsProcessorConfig `json:"processor"`
 	NotificationsEmailerConfig   NotificationsEmailerConfig   `json:"emailer"`
+	// Number of times to attempt recreating a notifications processor client should there be any disruptions.
+	ReconnectAttempts int `json:"reconnectAttempts"`
+	// Specifies the time interval to wait before attempting to reconnect the notifications processor client.
+	ReconnectDelaySeconds int `json:"reconnectDelaySeconds"`
```
Comment (wild-endeavor) [inline pkg/runtime/interfaces/application_configuration.go:155]: should this be int?  We can make it this if you want https://github.com/lyft/flytestdlib/blob/537f86093d86af270aab300742ee1a56f2905885/config/duration.go#L10

File: pkg/async/schedule/aws/workflow_executor.go (hunk 5 above)
Comment (wild-endeavor) [reply pkg/async/schedule/aws/workflow_executor.go:176]: and this is a layer above the async auto-retry layer right?  so if we exhaust all retries, then we chill for half an hour and then try again?

Comment (katrogan) [reply pkg/async/schedule/aws/workflow_executor.go:176]: wow half an hour is a goof, meant it to be half a minute - will update

File: pkg/runtime/interfaces/application_configuration.go (hunk 6 above)
Comment (katrogan) [reply pkg/runtime/interfaces/application_configuration.go:155]: sure, but it's a question of how much granularity we need to expose - are people really going to configure retry delays on the order or many minutes or even hours?

File: pkg/async/notifications/implementations/processor.go (hunk 4 above)
Comment (katrogan) [reply pkg/async/notifications/implementations/processor.go:48]: we already log a metric when the channel closes

File: pkg/async/schedule/aws/workflow_executor.go (hunk 5 above)
Comment (katrogan) [reply pkg/async/schedule/aws/workflow_executor.go:176]: added a metric on channel close

Comment (wild-endeavor) [reply pkg/async/schedule/aws/workflow_executor.go:177]: "async retry is for initializing clients a finite amount of times"
